src/nix-store/nix-store.cc

```diff
@@ -950,8 +950,12 @@ static void opServe(Strings opFlags, Strings opArgs)
                 info.sigs = readStrings<StringSet>(in);
                 in >> info.ca;
 
-                // FIXME: race if addToStore doesn't read source?
-                store->addToStore(info, in, NoRepair, NoCheckSigs);
+                SizedSource sizedSource(in, info.narSize);
```

Comment:
I don't really understand what `SizedSource` is for, in conjunction with `drainAll()`. Shouldn't `drainAll` just read and discard `remain` bytes?

Reply:
There should be a check for `info.narSize == 0`, which can happen in some legacy situations. It's probably fine to just throw an exception.

Reply:
The SizedSource is a way to track the `remain` bytes inside of `addToStore()` since we don't know how much it will consume. It also acts as a guard that `addToStore()` won't read more than `info.narSize` on the socket (not that this has been a problem).

Reply:
That's what I thought but I assumed the point is so that you can then skip `remain` bytes at the end (rather than drain the FD, which introduces problems when the client is pipelining requests or hasn't written everything yet).